### PR TITLE
[moe] support optimizer checkpoint

### PR DESCRIPTION
**Pull request by oahzxl**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [ ] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

* support optimizer checkpoint
* update test script

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [ ] I have performed a self-review of my code
- [ ] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/booster/plugin/moe_hybrid_parallel_plugin.py [hunk 1]
```diff
@@ -359,9 +359,6 @@ def configure(
                         max_norm=self.max_norm,
                         **self.amp_config,
                     )
-                    self.checkpoint_io.link_master_and_working_param(
```
Comment (ver217) [inline colossalai/booster/plugin/moe_hybrid_parallel_plugin.py:362]: How to keep the consistency of fp16 params and fp32 master params when loading model?

Comment (oahzxl) [reply colossalai/booster/plugin/moe_hybrid_parallel_plugin.py:362]: I align with hybrid parallel plugin in this change https://github.com/hpcaitech/ColossalAI/blob/c36e782d80bd3936fd3fec1f72dc0b476bf4016c/colossalai/booster/plugin/hybrid_parallel_plugin.py#L1093

File: tests/test_moe/test_moe_checkpoint.py [hunk 2]
```diff
@@ -6,57 +6,127 @@
...
+        "examples/language/openmoe",
+    )
+)
 
 OpenMoeForCausalLM = importlib.import_module("model.modeling_openmoe").OpenMoeForCausalLM
 set_openmoe_args = importlib.import_module("model.modeling_openmoe").set_openmoe_args
 OpenMoeForCausalLMPolicy = importlib.import_module("model.openmoe_policy").OpenMoeForCausalLMPolicy
 
 
+class RandomDataset(Dataset):
```
Comment (ver217) [inline tests/test_moe/test_moe_checkpoint.py:31]: Maybe you  can use `colossalai.testing.DummyDataloader`

File: tests/test_moe/test_moe_checkpoint.py [hunk 3]
```diff
@@ -6,57 +6,127 @@
...
+            "attention_mask": self.attention_mask[idx],
+            "labels": self.input_ids[idx],
+        }
+
+
+def run_fwd_bwd(
+    model, data, label, criterion, optimizer, enable_autocast=False, pipeline=False, booster=None, plugin=None
+):
+    model.train()
+    with torch.cuda.amp.autocast(enabled=enable_autocast):
```
Comment (ver217) [inline tests/test_moe/test_moe_checkpoint.py:53]: Zero plugin implements internal AMP. Autocast is not needed

File: tests/test_moe/test_moe_checkpoint.py [hunk 4]
```diff
@@ -66,51 +136,83 @@ def get_model(parallel):
...
-            seed=42,
             parallel="EP",
             mode="fixed",
             fixed_dp_size=1,
             fixed_ep_size=2,
             fixed_pp_size=2,
         )
-    model1, booster1 = get_model(parallel)
-    model2, booster2 = get_model(parallel)
+    model1, booster1, optim1 = get_model(parallel)
```
Comment (ver217) [inline tests/test_moe/test_moe_checkpoint.py:172]: What's the difference of model1 and model2?

Comment (oahzxl) [reply tests/test_moe/test_moe_checkpoint.py:172]: Model1's optim will be updated and model2 will load model1's optim

File: tests/test_moe/test_moe_checkpoint.py (hunk 2 above)
Comment (oahzxl) [reply tests/test_moe/test_moe_checkpoint.py:31]: done

File: tests/test_moe/test_moe_checkpoint.py (hunk 3 above)
Comment (oahzxl) [reply tests/test_moe/test_moe_checkpoint.py]: done

**Review by ver217 — APPROVED** [review]

**Review by binmakeswell — APPROVED** [review]

**Pull request merged**